### PR TITLE
Kubernetes Calico TLS

### PR DESCRIPTION
**Pull request by gambol99**
The current implementation when Etcd TLS was added does not support using calico as the configuration and client certificates are not present. This PR updates the calico manifests and adds the distribution of the client certificate

### CODE REVIEW

**Comment by gambol99** [comment]
can i have a quick review please @KashifSaadat

File: nodeup/pkg/model/calico.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+)
+
+// CalicoBuilder configures the calico CNI provider
+type CalicoBuilder struct {
+	*NodeupModelContext
+}
+
+var _ fi.ModelBuilder = &CalicoBuilder{}
+
+// Build is responsible for performing any setup to the calica CNI provider
```
Comment (KashifSaadat) [inline nodeup/pkg/model/calico.go]: `calica` => `calico`

File: nodeup/pkg/model/context.go [hunk 2]
```diff
@@ -268,3 +269,55 @@ func (c *NodeupModelContext) KubectlPath() string {
...
+		Contents: fi.NewStringResource(serialized),
+		Type:     nodetasks.FileType_File,
+		Mode:     s("0400"),
+	})
+
+	return nil
+}
+
+// buildPrivateKeyTask is responsible for build a certificate request task
+func (c *NodeupModelContext) BuildPrivateTask(ctx *fi.ModelBuilderContext, name, filename string) error {
```
Comment (KashifSaadat) [inline nodeup/pkg/model/context.go]: Function name differs from associated comment, just a minor correction to comment

File: pkg/model/iam/iam_builder.go [hunk 3]
```diff
@@ -219,10 +219,6 @@ func (b *PolicyBuilder) BuildAWSPolicyNode() (*Policy, error) {
 		addECRPermissions(p)
 	}
 
-	if b.Cluster.Spec.Networking != nil && b.Cluster.Spec.Networking.AmazonVPC != nil {
-		addAmazonVPCCNIPermissions(p, resource, b.Cluster.Spec.IAM.Legacy, b.Cluster.GetName())
```
Comment (KashifSaadat) [inline pkg/model/iam/iam_builder.go]: Did you mean to delete this? It should probably be included in the Networking Spec checks lower.

File: pkg/model/pki.go [hunk 4]
```diff
@@ -109,23 +109,30 @@ func (b *PKIModelBuilder) Build(c *fi.ModelBuilderContext) error {
...
-				Lifecycle:      b.Lifecycle,
-				Name:           fi.String("etcd"),
-				Subject:        "cn=etcd",
-				Type:           "server",
-				Signer:         defaultCA,
-			})
-		}
-		{
+		// @question should wildcard's be here instead of generating per node. If we ever provide the
+		// ability to resize the master, this will become a blocker
```
Comment (KashifSaadat) [inline pkg/model/pki.go]: Ping @chrislovecnm

File: upup/pkg/fi/cloudup/template_functions.go [hunk 5]
```diff
@@ -43,25 +43,24 @@ import (
...
 }
 
 // This will define the available functions we can use in our YAML models
 // If we are trying to get a new function implemented it MUST
 // be defined here.
 func (tf *TemplateFunctions) AddTo(dest template.FuncMap) {
+	dest["EtcdScheme"] = tf.EtcdScheme
 	dest["SharedVPC"] = tf.SharedVPC
-
+	dest["UserEtcdTLS"] = tf.UseEtcdTLS
```
Comment (KashifSaadat) [inline upup/pkg/fi/cloudup/template_functions.go]: Typo here? Should be `dest["UseEtcdTLS"] = tf.UseEtcdTLS`

**Comment by blakebarnett** [comment]
lgtm

**Comment by chrislovecnm** [comment]
/hold

Did anyone test upgrades?  Moving from TLS from non-tls?

**Comment by blakebarnett** [comment]
Testing this now, I'm getting:

```Jan 20 01:11:01 ip-10-88-1-10 nodeup[775]: I0120 01:11:01.536351     775 s3fs.go:176] Reading file "s3://kops-k8s-state-store/e2e.us-east-1.aws.k8s/pki/issued/master/keyset.yaml"
Jan 20 01:11:01 ip-10-88-1-10 nodeup[775]: W0120 01:11:01.546804     775 main.go:141] got error running nodeup (will retry in 30s): error building loader: error in 'FindCert' attempting to load cert "master": error loading certificate "s3://kops-k8s-state-store/e2e.us-east-1.aws.k8s/pki/issued/master/keyset.yaml": could not parse certificate```

**Comment by blakebarnett** [comment]
Didn't set KOPS_BASE_URL in my last test, redoing :)

**Comment by gambol99** [comment]
hi @chrislovecnm .. this has been test from a new build only ... 
> Moving from TLS from non-tls?

Given this would require a migration from non-tls etcd to tls etcd which we don't yet support it can't really be tested

**Comment by blakebarnett** [comment]
Verified this requires manual intervention and downtime.

**Comment by blakebarnett** [comment]
I'm getting this in my tests (verified it's downloading the build for nodeup):

```
Jan 22 19:01:58 ip-10-88-1-10 nodeup[761]: I0122 19:01:58.914308     761 s3fs.go:198] Reading file "s3://<bucket-name>/e2e.us-east-1.aws.k8s/pki/issued/ca/keyset.yaml"
Jan 22 19:01:58 ip-10-88-1-10 nodeup[761]: W0122 19:01:58.920944     761 main.go:141] got error running nodeup (will retry in 30s): error building loader: CA certificate "ca" not found
```

**Comment by blakebarnett** [comment]
ok, this appears to be a more general issue, I am still getting it on master without any of this branch.

**Comment by gambol99** [comment]
hi @blakebarnett .. this is related to https://github.com/kubernetes/kops/issues/4156 ... My tests work as it was a new cluster build, where as the https://github.com/kubernetes/kops/issues/4156 effects upgrading i believe .. @KashifSaadat??

**Comment by blakebarnett** [comment]
*sigh*, wish I'd seen that issue before.... so much time wasted, heh.

**Comment by KashifSaadat** [comment]
Yep that's correct. #4156 highlights two issues, one of which is that you receive errors such as above on a cluster upgrade if the keyset files don't exist in S3 (they are only created on new cluster creation).

**Comment by chrislovecnm** [comment]
@KashifSaadat 

> one of which is that you receive errors such as above on a cluster upgrade if the keyset files don't exist in S3

More context, please.  Probably should open an issue to discuss this

**Comment by KashifSaadat** [comment]
Hey @chrislovecnm, it was partially mentioned in #4156 but I've raised a separate issue to track this now: #4351

**Comment by chrislovecnm** [comment]
@gambol99 this still has a `/hold` on the PR, are we good to go?

**Comment by gambol99** [comment]
hi @chrislovecnm ... rebased and retested as working :-) .. good to go

**Comment by chrislovecnm** [comment]
@gambol99 we are going to need release notes on this one :)

/hold cancel
/lgtm

**Pull request merged**